proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+Decorate exceptions with backtrace information
+==============================================
+
+.. authors:: Ben Gamari; David Eichmann; Sven Tennie
+.. date-accepted::
+.. ticket-url:: https://gitlab.haskell.org/ghc/ghc/issues/18159
+.. implemented:: in-progress <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8869>
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/330>`_.
+.. contents::
+
+Ease localization of runtime errors reported via the synchronous exception mechanism
+by attaching backtraces to exceptions.
+
+
+Motivation
+----------
+Exceptions are one of the primary mechanisms by which Haskell programs report
+errors. However, in contrast to most languages, Haskell provides only few tools for
+identifying the source of such errors. This leads to a poor debugging experience
+and makes it difficult to monitor systems in production.
+
+While over the past few years GHC has grown a variety of mechanisms for reporting
+backtraces (e.g. ``HasCallStack``, ``GHC.Stack.CCS``, and DWARF debug
+information), currently we do not have a means to attach such backtraces to
+exceptions. The goal of this proposal is to fix this `long-standing
+<https://www.youtube.com/watch?v=J0c4L-AURDQ>`_ problem.
+
+We want to ensure that exceptions report provenance information *by
+default* without requiring action on the part of the developer. To provide this provenance we leverage
+the existing mechanisms for collecting backtraces listed above. Furthermore, we
+want to ensure that this information is available for consumption in structured
+form by the user program, to allow use by logging libraries (for instance, see
+`kaptip-raven #1
+<https://github.com/cachix/katip-raven/issues/1#issuecomment-625389463>`_),
+automatic error reporting, code analysis tools, and the like.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal consists of three largely-independent parts:
+
+1. augmenting the current root of the exception hierarchy,
+   ``SomeException``, with additional metadata in the form of an
+   ``ExceptionContext``.
+2. the introduction of annotation types to capture backtraces from the
+   above-described backtrace-collection mechanisms
+3. a facility for choosing which backtrace mechanism(s)
+   should be used to collect exception provenance.
```

Comment:
```suggestion
This proposal consists of four largely-independent parts:

1. augmenting the current root of the exception hierarchy,
   ``SomeException``, with additional metadata in the form of an
   ``ExceptionContext``;
2. the introduction of annotation types to capture backtraces from the
   above-described backtrace-collection mechanisms;
3. a facility for choosing which backtrace mechanism(s)
   should be used to collect exception provenance;
4. changing the top level exception handler to use `displayException`
   rather than `show`.
```